crates/driver/src/domain/competition/quote.rs

```diff
@@ -0,0 +1,182 @@
+            competition::order::Side::Buy => Self {
+                sell: fulfillment.order.sell,
+                buy: fulfillment.executed.to_asset(&fulfillment.order),
+                interactions,
+            },
+            competition::order::Side::Sell => Self {
+                sell: fulfillment.executed.to_asset(&fulfillment.order),
+                buy: fulfillment.order.buy,
+                interactions,
+            },
+        }
+    }
+}
+
+/// An order which needs to be quoted.
+#[derive(Debug)]
+pub struct Order {
+    pub sell_token: eth::TokenAddress,
+    pub buy_token: eth::TokenAddress,
+    pub amount: Amount,
+    pub valid_to: u32,
+    /// See [`crate::domain::competition::order::Partial`].
+    pub partial: bool,
+    pub quality: Quality,
+    pub gas_price: eth::EffectiveGasPrice,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum Amount {
+    Sell(eth::U256),
+    Buy(eth::U256),
+}
+
+impl Order {
+    /// The asset being bought. Zero if this is a sell.
+    pub fn buy(&self) -> eth::Asset {
+        match self.amount {
+            Amount::Sell(..) => eth::Asset {
+                amount: Default::default(),
+                token: self.buy_token,
+            },
+            Amount::Buy(amount) => eth::Asset {
+                amount,
+                token: self.buy_token,
+            },
+        }
+    }
+
+    /// The asset being sold. Zero if this is a buy.
+    pub fn sell(&self) -> eth::Asset {
+        match self.amount {
+            Amount::Sell(amount) => eth::Asset {
+                amount,
+                token: self.sell_token,
+            },
+            Amount::Buy(..) => eth::Asset {
+                amount: Default::default(),
+                token: self.sell_token,
+            },
+        }
+    }
+
+    pub fn side(&self) -> competition::order::Side {
+        match self.amount {
+            Amount::Sell(..) => competition::order::Side::Sell,
+            Amount::Buy(..) => competition::order::Side::Buy,
+        }
+    }
+}
+
+/// Quality of the quote to be generated. This value determines the time
+/// allocated for the solver to generate the solution.
+#[derive(Debug, Clone, Copy)]
+pub enum Quality {
+    Fast,
+    Optimal,
+}
+
+impl Order {
+    /// Generate a quote for this order. This calls `/solve` on the solver with
+    /// a "fake" auction which contains a single order, and then determines
+    /// the quote for the order based on the solution that the solver
+    /// returns.
+    pub async fn quote(&self, solver: &Solver, config: &Config) -> Result<Quote, Error> {
+        let solution = solver
+            .solve(&self.fake_auction(), self.timeout(config))
+            .await?;
+        let fulfillment = solution
+            .trades
+            .into_iter()
+            .filter_map(|trade| match trade {
+                competition::solution::Trade::Fulfillment(fulfillment) => Some(fulfillment),
+                competition::solution::Trade::Jit(..) => None,
```

Comment:
Not sure if we should disallow solvers quoting an order based on a `Jit` order. I'm not sure but possibly a `Jit` order would cause no problems during quote verification (i.e. trade simulation).

Reply:
JIT orders should be supported. Unfortunately, they cannot be modeled with `Interaction`, so it would require adding an additional field to the `/quote` API's result type for JIT orders specifically.

Reply:
> Also the ordering between JIT orders and fulfillments matters, right?

No, it shouldn't.